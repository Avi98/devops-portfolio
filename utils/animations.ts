import { keyframes } from "@/theme/stitches.config"

export const reboundAnimation = keyframes({
  "0%": {
    transform: "scale(0.95)",
  },
  "40%": {
    transform: "scale(1.02)",
  },
  "80%": {
    transform: "scale(0.98)",
  },
  "100%": {
    transform: "scale(1)",
  },
})

export const appearanceIn = keyframes({
  "0%": {
    opacity: 0,
    transform: "scale(0.95)",
  },
  "60%": {
    opacity: 0.75,
    transform: "scale(1.02)",
  },
  "100%": {
    opacity: 1,
    transform: "scale(1)",
  },
})

export const appearanceOut = keyframes({
  "0%": {
    opacity: 1,
    transform: "scale(1)",
  },
  "100%": {
    opacity: 0,
    transform: "scale(0.95)",
  },
})

export const levitating = keyframes({
  "0%": {
    transform: "translateY(0)",
  },
  "30%": {
    transform: "translateY(-10px)",
  },
  "50%": {
    transform: "translateY(4px)",
  },
  "70%": {
    transform: "translateY(-15px)",
  },
  "100%": {
    transform: "translateY(0)",
  },
})

export const appears = keyframes({
  from: {
    opacity: 0,
  },
  to: {
    opacity: 1,
  },
})

export const pulse = keyframes({
  "0%": {
    transform: "scale(1)",
  },
  "50%": {
    transform: "scale(1.2)",
  },
  "100%": {
    transform: "scale(1)",
  },
})
