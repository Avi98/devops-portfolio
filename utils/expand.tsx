import React, { useEffect, useRef, useState } from "react"

import { CSS, styled } from "@/theme/stitches.config"
import withDefaults from "@/utils/with-defaults"

import useRealShape from "./use-real-shape"

interface Props {
  isExpanded?: boolean
  animated?: boolean
  delay?: number
  css?: CSS
}

const defaultProps = {
  isExpanded: false,
  animated: true,
  delay: 200,
}

const StyledExpand = styled("div", {
  p: 0,
  m: 0,
  h: 0,
  opacity: 0,
  overflow: "hidden",
  variants: {
    expanded: {
      true: {
        opacity: 1,
      },
    },
  },
})

export type ExpandProps = Props & typeof defaultProps

const Expand: React.FC<React.PropsWithChildren<ExpandProps>> = ({
  isExpanded,
  delay,
  animated,
  css,
  children,
}) => {
  const [height, setHeight] = useState<string>(isExpanded ? "auto" : "0")
  const [selfExpanded, setSelfExpanded] = useState<boolean>(isExpanded)
  const contentRef = useRef<HTMLDivElement>(null)
  const entryTimer = useRef<number>()
  const leaveTimer = useRef<number>()
  const resetTimer = useRef<number>()
  const [state, updateShape] = useRealShape<HTMLDivElement>(contentRef)

  useEffect(() => setHeight(`${state.height}px`), [state.height])
  useEffect(() => {
    if (isExpanded === selfExpanded) return
    // show element or reset height.
    // force an update once manually, even if the element does not change.
    // (the height of the element might be "auto")
    if (!isExpanded) {
      updateShape()
      setHeight(`${state.height}px`)
    }

    // show expand animation
    entryTimer.current = window.setTimeout(() => {
      setSelfExpanded(isExpanded)
      clearTimeout(entryTimer.current)
    }, 30)

    // Reset height after animation
    if (isExpanded) {
      resetTimer.current = window.setTimeout(() => {
        setHeight("auto")
        clearTimeout(resetTimer.current)
      }, delay)
    } else {
      leaveTimer.current = window.setTimeout(() => {
        clearTimeout(leaveTimer.current)
      }, delay / 2)
    }

    return () => {
      clearTimeout(entryTimer.current)
      clearTimeout(leaveTimer.current)
      clearTimeout(resetTimer.current)
    }
  }, [isExpanded])

  return (
    <StyledExpand
      css={{
        height: selfExpanded ? height : "0",
        transition: animated
          ? `height ${delay}ms ease 0ms,
    opacity ${delay * 1.5}ms ease 0ms;`
          : "none",
        ...(css as any),
      }}
      expanded={selfExpanded}
    >
      <div ref={contentRef} className="devui-expand-content">
        {children}
      </div>
    </StyledExpand>
  )
}

export default withDefaults(Expand, defaultProps)
