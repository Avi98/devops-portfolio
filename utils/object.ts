import { Key } from "react"

export const isObject = (target: unknown) =>
  target && typeof target === "object"

export interface MergeObject {
  [key: string]: any
}

export const renameProp = (
  oldProp: string,
  newProp: string,
  { [oldProp]: old, ...others }
) => ({
  [newProp]: old,
  ...others,
})

export const copyObject = (obj: any) => {
  if (!isObject(obj)) return obj
  if (obj instanceof Array) return [...obj]
  return { ...obj }
}

export const omitObject = (obj: any, omitKeys: string[]) => {
  if (!isObject(obj)) return obj
  if (obj instanceof Array) return [...obj]
  const newObj = { ...obj }
  omitKeys.forEach(key => newObj[key] && delete newObj[key])
  return newObj
}

export const cleanObject = (obj: any) => {
  if (!isObject(obj)) return obj
  if (obj instanceof Array) return [...obj]
  const newObj = { ...obj }
  Object.keys(newObj).forEach(key => {
    if (newObj[key] === undefined || newObj[key] === null) {
      delete newObj[key]
    }
  })
  return newObj
}

export const cleanObjectKeys = (obj: any, keys: string[] = []) => {
  if (!isObject(obj)) return obj
  if (obj instanceof Array) return [...obj]
  const newObj = { ...obj }
  keys.forEach(key => {
    if (newObj[key]) {
      delete newObj[key]
    }
  })
  return newObj
}

export const getKeyValue = (obj: any, key: Key) => {
  if (!isObject(obj)) return obj
  if (obj instanceof Array) return [...obj]
  return obj[key]
}
