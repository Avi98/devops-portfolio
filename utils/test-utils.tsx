import { FC, ReactElement, ReactNode } from "react"

import { render, RenderOptions } from "@testing-library/react"

import MemoCssBaseline from "@/styles/css-baseline"
import { DevOpsInterface } from "@/theme"

const AllTheProviders: FC<{ children: ReactNode }> = ({ children }) => {
  return (
    <DevOpsInterface>
      <MemoCssBaseline />
      {children}
    </DevOpsInterface>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options })

export * from "@testing-library/react"
export { customRender as render }
