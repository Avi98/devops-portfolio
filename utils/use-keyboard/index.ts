import { KeyCode, KeyMod } from "./codes"
import useKeyboard from "./use-keyboard"

export { KeyMod, KeyCode }

export type {
  UseKeyboardHandler,
  KeyboardOptions,
  KeyboardResult,
  UseKeyboard,
} from "./use-keyboard"

export default useKeyboard
