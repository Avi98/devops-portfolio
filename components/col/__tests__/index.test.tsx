import React from "react"

import { render, screen } from "@/utils/test-utils"

import Col from "../index"

describe("Col", () => {
  it("should render column correctly", () => {
    render(<Col>col</Col>)
    expect(screen).toMatchSnapshot()
  })

  it("should render column correctly when nested", () => {
    render(
      <Col>
        <Col>
          <Col />
          col
        </Col>
      </Col>
    )
    expect(screen).toMatchSnapshot()
  })
})
