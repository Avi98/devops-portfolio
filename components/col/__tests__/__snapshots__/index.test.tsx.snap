// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Col should render column correctly 1`] = `
{
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`Col should render column correctly when nested 1`] = `
{
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`Col should work with span and offset 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-overlay-container="true"
      >
        <div>
          <div
            class="devopsui-c-cMTZRp devopsui-c-cMTZRp-ikSMVmn-css"
          >
            col
          </div>
          <div
            class="devopsui-c-cMTZRp devopsui-c-cMTZRp-icjwvEK-css"
          >
            col
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-overlay-container="true"
    >
      <div>
        <div
          class="devopsui-c-cMTZRp devopsui-c-cMTZRp-ikSMVmn-css"
        >
          col
        </div>
        <div
          class="devopsui-c-cMTZRp devopsui-c-cMTZRp-icjwvEK-css"
        >
          col
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
